### PR TITLE
Change set typeof check to allow functions

### PR DESCRIPTION
**Pull request by thoov**
Native WeakMaps allow:

```js
var wm = new WeakMap();
wm.set(function() {}, 'whyyyyyyy');
```

so updating the logic to support this

### CODE REVIEW

File: packages/ember-metal/tests/weak_map_test.js [hunk 1]
```diff
@@ -53,6 +53,26 @@ QUnit.test('that error is thrown when using a non object key', function(assert)
   throws(function() {
     map.set('a', 1);
   }, /Uncaught TypeError: Invalid value used as weak map key/);
+
+  throws(function() {
```
Comment (rwjblue) [inline packages/ember-metal/tests/weak_map_test.js]: assertion testing should use `expectAssertion` (`assert` does not throw in prod builds so `expectAssertion` handles that for us).

Comment (thoov) [inline packages/ember-metal/tests/weak_map_test.js]: thanks @rwjblue. did not know this. I will update

File: packages/ember-metal/lib/weak_map.js [hunk 2]
```diff
@@ -41,7 +54,7 @@ WeakMap.prototype.get = function(obj) {
  * @return {Any} stored value
  */
 WeakMap.prototype.set = function(obj, value) {
-  assert('Uncaught TypeError: Invalid value used as weak map key', typeOf(obj) === 'object');
+  assert('Uncaught TypeError: Invalid value used as weak map key', !isPrimitiveType(obj));
```
Comment (stefanpenner) [inline packages/ember-metal/lib/weak_map.js]: is this a whitelist of `object` + `function` ? or is it more.

Comment (thoov) [inline packages/ember-metal/lib/weak_map.js]: @stefanpenner yes that is all its doing. Do you prefer just checking instead?

```js
typeof obj === 'object' || typeof obj === 'function'
```

Comment (stefanpenner) [inline packages/ember-metal/lib/weak_map.js]: Well, this is a black-list. I would prefer a white-list

Comment (stefanpenner) [inline packages/ember-metal/lib/weak_map.js]: is `null` allowed?

Comment (thoov) [inline packages/ember-metal/lib/weak_map.js]: @stefanpenner moved to a whitelist and no null should not be allowed as a key

Comment (stefanpenner) [inline packages/ember-metal/lib/weak_map.js]: @thoov but `typeof null === 'object'`

Comment (thoov) [inline packages/ember-metal/lib/weak_map.js]: @stefanpenner o right... that stupid thing :-) fixed

Comment (stefanpenner) [inline packages/ember-metal/lib/weak_map.js]: ya its tricky

**Comment by stefanpenner** [comment]
awesome, thanks dude!

**Pull request merged**